### PR TITLE
New Feature: jwt authn stream filter

### PR DESCRIPTION
**Pull request by GLYASAI**
Currently supports istio official task [Authorization with JWT](https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/).

Example config:
```
"stream_filters": [
            {
              "type": "jwt_authn",
              "config": {
                "providers": {
                  "origins-0": {
                    "issuer": "testing@secure.istio.io",
                    "localJwks": {
                      "inlineString": "{ \"keys\":[ {\"e\":\"AQAB\",\"kid\":\"DHFbpoIUqrY8t2zpA2qXfCmr5VO5ZEr4RzHU_-envvQ\",\"kty\":\"RSA\",\"n\":\"xAE7eB6qugXyCAG3yhh7pkDkT65pHymX-P7KfIupjf59vsdo91bSP9C8H07pSAGQO1MV_xFj9VswgsCg4R6otmg5PV2He95lZdHtOcU5DXIg_pbhLdKXbi66GlVeK6ABZOUW3WYtnNHD-91gVuoeJT_DwtGGcp4ignkgXfkiEm4sw-4sfb4qdt5oLbyVpmW6x9cfa7vs2WTfURiCrBoUqgBo_-4WTiULmmHSGZHOjzwa8WtrtOQGsAFjIbno85jp6MnGGGZPYZbDAa_b3y5u-YpW7ypZrvD8BgtKVjgtQgZhLAGezMt0ua3DRrWnKqTZ0BJ_EyxOGuHJrLsn00fnMQ\"}]}"
                    }
                  }
                },
                "rules": [
                  {
                    "match": {
                      "prefix": "/"
                    },
                    "requires": {
                      "requiresAny": {
                        "requirements": [
                          {
                            "providerName": "origins-0"
                          },
                          {
                            "allowMissing": {}
                          }
                        ]
                      }
                    }
                  }
                ]
              }
            }
          ]
        }
```

### Solutions

Use rules to match user requests, and then use jwks in the provider to verify the token in the user request header

### CODE REVIEW

File: pkg/filter/stream/jwtauthn/filter.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+func (f *filter) OnReceive(ctx context.Context, headers api.HeaderMap, buf buffer.IoBuffer, trailers api.HeaderMap) api.StreamFilterStatus {
+	log.DefaultLogger.Debugf("call jwt_authn filter")
+
+	// TODO(huangrh): bypass
+
+	requestArg, err := variable.GetVariableValue(ctx, types.VarHttpRequestArg)
+	if err != nil {
+		log.DefaultLogger.Errorf("[jwt_authn filter] get query parameter: %v", err)
+		return api.StreamFilterStop
```
Comment (wangfakang) [inline pkg/filter/stream/jwtauthn/filter.go:31]: Not all requests have request parameters, so i think `api.StreamFilterContinue` should be returned

**Comment by wangfakang** [comment]
@GLYASAI You could add an example to  `examples` dir.

File: pkg/filter/stream/jwtauthn/authenticator.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	provider                string
+	jwksCache               JwksCache
+	jwksData                JwksData
+	allowFail, allowMissing bool
+	fetcher                 JwksFetcher
+
+	tokens     []JwtLocation
+	currToken  JwtLocation
+	headers    api.HeaderMap
+	jwt        *jwt.Token
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/authenticator.go:29]: reviewdog ：`jwt` is unused

File: pkg/filter/stream/jwtauthn/authenticator.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+		provider:     provider,
+		jwksCache:    jwksCache,
+		fetcher:      fetcher,
+		allowFail:    allowFail,
+		allowMissing: allowMissing,
+	}
+}
+
+func (a *authenticator) Verify(headers api.HeaderMap, tokens []JwtLocation) error {
+	log.DefaultLogger.Infof("JWT authentication starts, tokens size=%d", len(tokens))
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/authenticator.go:44]: judge log level

File: pkg/filter/stream/jwtauthn/authenticator.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+		return "_IS_ALLOW_FALED_"
+	}
+	if a.allowMissing {
+		return "_IS_ALLOW_MISSING_"
+	}
+	return "_UNKNOWN_"
+}
+
+func (a *authenticator) done(err error) error {
+	log.DefaultLogger.Debugf("%s: JWT token verification completed with: %v", a.name(), err)
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/authenticator.go:73]: judge log level

File: pkg/filter/stream/jwtauthn/authenticator.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+	// if on allow missing or failed this should verify all tokens, otherwise stop on ok.
+	if len(a.tokens) == 0 {
+		a.tokens = nil // clear tokens
+		if a.allowFail {
+			return nil
+		} else if a.allowMissing && err == ErrJwtNotFound {
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/authenticator.go:80]: else if  -> if

File: pkg/filter/stream/jwtauthn/authenticator.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+	return a.startVerify()
+}
+
+func (a *authenticator) startVerify() error {
+	a.currToken = a.tokens[0]
+	a.tokens = a.tokens[1:]
+
+	jwtToken, parts, err := new(jwt.Parser).ParseUnverified(a.currToken.Token(), &jwt.StandardClaims{})
+	if err != nil {
+		return a.done(ErrJwtBadFormat)
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/authenticator.go:99]: 这儿不会死循环吗？  a.done 会继续调用 startVerify

File: pkg/filter/stream/jwtauthn/authenticator.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	a.jwtPayload = parts[1]
+	claims := jwtToken.Claims.(*jwt.StandardClaims)
+
+	log.DefaultLogger.Debugf("Verifying JWT token of issuer: %s", claims.Issuer)
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/authenticator.go:104]: judge log level

File: pkg/filter/stream/jwtauthn/authenticator.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+func (a *authenticator) verifyKey() error {
+	_, err := jwt.Parse(a.currToken.Token(), func(token *jwt.Token) (interface{}, error) {
+		kid, ok := token.Header["kid"].(string)
+		if !ok {
+			return nil, errors.New("expecting JWT header to have string kid")
+		}
+
+		keys := a.jwksData.GetJwksObj().LookupKeyID(kid)
+		if len(keys) == 0 {
+			return nil, fmt.Errorf("unable to find key %q", kid)
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/authenticator.go:165]: %q - > %s

File: pkg/filter/stream/jwtauthn/extractor.go [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+	jwtauthnv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/jwt_authn/v3"
+	"github.com/valyala/fasthttp"
+	"mosn.io/api"
+)
+
+const (
+	//
+	authorization = "Authorization"
+	// The header value prefix for Authorization.
+	bearerPrefix = "Bearer "
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/extractor.go:15]: header是不能有空格的哟？

File: pkg/filter/stream/jwtauthn/extractor.go [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+func createHeaderLocationSpec(headerName, valuePrefix string) *HeaderLocationSpec {
+	return &HeaderLocationSpec{
+		headerName:       headerName,
+		valuePrefix:      valuePrefix,
+		specifiedIssuers: make(map[string]struct{}),
+	}
+}
+
+func (h *HeaderLocationSpec) addIssuer(issuer string) {
+	h.specifiedIssuers[issuer] = struct{}{}
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/extractor.go:56]: 会有并发问题吗？

File: pkg/filter/stream/jwtauthn/extractor.go [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+func createParamLocationSpec() *ParamLocationSpec {
+	return &ParamLocationSpec{
+		specifiedIssuers: make(map[string]struct{}),
+	}
+}
+
+func (p *ParamLocationSpec) addIssuer(issuer string) {
+	p.specifiedIssuers[issuer] = struct{}{}
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/extractor.go:72]: 同上

File: pkg/filter/stream/jwtauthn/factory.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// FilterConfigFactory -
+type FilterConfigFactory struct {
+	config *jwtauthnv3.JwtAuthentication
+}
+
+// CreateJwtAuthnFilterFactory creates a new JwtAuthnFilterFactory.
+func CreateJwtAuthnFilterFactory(cfg map[string]interface{}) (api.StreamFilterChainFactory, error) {
+	log.DefaultLogger.Debugf("create jwt authn stream filter factory")
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/factory.go:27]: judge log level

File: pkg/filter/stream/jwtauthn/filter.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+func newJwtAuthnFilter(config FilterConfig) *filter {
+	return &filter{
+		config: config,
+	}
+}
+
+func (f *filter) OnReceive(ctx context.Context, headers api.HeaderMap, buf buffer.IoBuffer, trailers api.HeaderMap) api.StreamFilterStatus {
+	log.DefaultLogger.Debugf("call jwt_authn filter")
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/filter.go:24]: judge log level

File: pkg/filter/stream/jwtauthn/header_matcher.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+	"mosn.io/api"
+)
+
+func matchHeaders(requestHeaders api.HeaderMap, configHeaders []*routev3.HeaderMatcher) bool {
+	// No headers to match is considered a match.
+	if len(configHeaders) == 0 {
+		return true
+	}
+
+	if requestHeaders == nil {
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/header_matcher.go:16]: maybe requestHeaders len is 0

File: pkg/filter/stream/jwtauthn/matcher.go [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+		headers:         rule.GetMatch().GetHeaders(),
+		queryParameters: rule.GetMatch().GetQueryParameters(),
+	}
+}
+
+// Check match for HeaderMatcher and QueryParameterMatcher
+func (p *baseMatcher) matchRoutes(headers api.HeaderMap, requestPath string) bool {
+	matchs := matchHeaders(headers, p.headers)
+	if len(p.queryParameters) > 0 {
+
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/matcher.go:54]: ?

File: pkg/filter/stream/jwtauthn/matcher.go [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+	baseMatcher := newBaseMatcher(rule)
+	return &prefixMatcher{
+		prefix:      rule.GetMatch().GetPrefix(),
+		baseMatcher: baseMatcher,
+	}
+}
+
+func (p *prefixMatcher) Matches(headers api.HeaderMap, requestPath string) bool {
+	if p.baseMatcher.matchRoutes(headers, requestPath) && strings.HasPrefix(requestPath, p.prefix) {
+		log.DefaultLogger.Debugf("Prefix requirement '%s' matched.", p.prefix)
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/matcher.go:74]: judge log level

File: pkg/filter/stream/jwtauthn/matcher.go [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+func (p *pathMatcher) Matches(headers api.HeaderMap, requestPath string) bool {
+	u, _ := url.Parse(requestPath)
+	requestPath = u.Path
+
+	pathMatch := p.path == requestPath
+	if !p.caseSensitive {
+		pathMatch = strings.ToLower(p.path) == strings.ToLower(requestPath)
+	}
+	if p.baseMatcher.matchRoutes(headers, requestPath) && pathMatch {
+		log.DefaultLogger.Debugf("Path requirement '%s' matched.", p.path)
```
Comment (taoyuanyuan) [inline pkg/filter/stream/jwtauthn/matcher.go:104]: judge log level

File: pkg/filter/stream/jwtauthn/matcher.go (hunk 15 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/matcher.go]: At present, only extracting token from headers is implemented, and extracting from query parameter has not been implemented yet

File: pkg/filter/stream/jwtauthn/extractor.go (hunk 9 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/extractor.go:15]: `Authorization: Bearer xxx`, Bearer 是 token 的前缀哦

File: pkg/filter/stream/jwtauthn/extractor.go (hunk 10 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/extractor.go:56]: 没有的. addIssuer 只会在初始化 Extractor 的时候调用. CreateFilterChain 创建新的 JwtFilter 的时候才会初始化 Extractor, 解析配置文件里面的 issuer, 之后就不会再 addIssuer 了.

File: pkg/filter/stream/jwtauthn/header_matcher.go (hunk 14 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/header_matcher.go:16]: 好像没有办法获取到 requestHeaders 的长度. 不判断长度也没有关系的, len 为 0 的话是过不了这里的 https://github.com/GLYASAI/mosn/blob/98a43c4c0550eabda6c3b42539ca6c3a3041010c/pkg/filter/stream/jwtauthn/header_matcher.go#L30

File: pkg/filter/stream/jwtauthn/filter.go (hunk 1 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/filter.go]: ok

File: pkg/filter/stream/jwtauthn/authenticator.go (hunk 2 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/authenticator.go]: ok

File: pkg/filter/stream/jwtauthn/authenticator.go (hunk 3 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/authenticator.go]: ok

File: pkg/filter/stream/jwtauthn/authenticator.go (hunk 4 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/authenticator.go]: ok

File: pkg/filter/stream/jwtauthn/authenticator.go (hunk 5 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/authenticator.go]: ok

File: pkg/filter/stream/jwtauthn/authenticator.go (hunk 7 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/authenticator.go]: ok

File: pkg/filter/stream/jwtauthn/authenticator.go (hunk 8 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/authenticator.go]: ok

File: pkg/filter/stream/jwtauthn/factory.go (hunk 12 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/factory.go]: ok

File: pkg/filter/stream/jwtauthn/matcher.go (hunk 16 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/matcher.go]: ok

File: pkg/filter/stream/jwtauthn/matcher.go (hunk 17 above)
Comment (GLYASAI) [reply pkg/filter/stream/jwtauthn/matcher.go]: ok

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Comment by wangfakang** [comment]
> @GLYASAI You could add an example to `examples` dir.

@GLYASAI  https://github.com/mosn/mosn/tree/master/examples/cn_readme

**Comment by GLYASAI** [comment]
@wangfakang Example added

**Review by wangfakang — APPROVED** [review]
LGTM.

**Pull request merged**